lib/authentication.js

```diff
@@ -0,0 +1,220 @@
+  ],
+  ONE_LOWERCASE_CHAR: [
+    password => !LOWERCASE_REGEXP.test(password),
+    'The password must contain lowercase char'
+  ],
+  ONE_UPPERERCASE_CHAR: [
+    password => !UPPERCASE_REGEXP.test(password),
+    'The password must contain uppercase char'
+  ],
+  ONE_NUMBER: [
+    password => !/[0-9]/.test(password),
+    'The password must contain number'
+  ],
+  ONE_SPECIAL_CHAR: [
+    password => !/[^A-Za-z0-9]/.test(password),
+    'The password must contain special char'
+  ]
+};
+
+const loginTests = {
+  MIN_LENGTH: [
+    login => login.length < MIN_LOGIN_LENGTH,
+    `The login must be longer than ${MIN_LOGIN_LENGTH} characters`
+  ],
+  MAX_LENGTH: [
+    login => login.length > MAX_LOGIN_LENGTH,
+    `The login must be less than ${MAX_LOGIN_LENGTH} characters`
+  ],
+  EMAIL_TEST: [
+    login => !EMAIL_REGEXP.test(login),
+    'Login must be email address'
+  ]
+};
+
+const loginPasswordTests = {
+  LOGIN_INCLUDES_PASSWORD: [
+    (login, password) => login.includes(password),
+    'The login can not contain a password'
+  ],
+  PASSWORD_INCLUDES_LOGIN: [
+    (login, password) => password.includes(login),
+    'The password can not contain a login'
+  ]
+};
+
+class AuthenticationStrength {
+  constructor(
+    hints, // Array, [{ required: Array, optional: Array }, ...]
+    compliance // Number, percent of best test group
+  ) {
+    this.hints = hints;
+    this.compliance = compliance;
+  }
+  get valid() {
+    return !!this.compliance;
+  }
+  get strength() {
+    if (this.compliance < 25) {
+      return 'Very weak';
+    } else if (this.compliance < 40) {
+      return 'Weak';
+    } else if (this.compliance < 60) {
+      return 'Good';
+    } else if (this.compliance < 80) {
+      return 'Strong';
+    } else {
+      return 'Very strong';
+    }
+  }
+}
+
+const makeTest = (
+  // Function that checks the arguments on a test suite
+  tests, // Array, array of password/login tests
+  testGroups, // Array, [{ required: [testNames], optional: [testNames] }]
+  ...testArgs // Array, [password] / [login] / [login, password]
+  // Returns: AuthenticationStrength instance
+) => {
+  const test = testsNames => {
+    const testsHints = [];
+    testsNames.forEach(testName => {
+      const [test, hint] = tests[testName];
+      if (test(...testArgs)) testsHints.push(hint);
+    });
+    return testsHints;
+  };
+
+  const hints = [];
+  let compliance = 0;
+
+  testGroups.forEach(({ required, optional }) => {
+    const requiredHints = test(required);
+    const optionalHints = test(optional);
+
+    if (!requiredHints.length) {
+      const optTestsCount = optional.length;
+      const optHintsCount = optionalHints.length;
+      const percent = 100 * (optTestsCount - optHintsCount) / optTestsCount;
+      compliance = Math.max(compliance, percent);
+    }
+
+    hints.push({ required: requiredHints, optional: optionalHints });
+  });
+
+  hints.sort((prevHints, hints) => {
```

Comment:
What's this sorting for? The only way for the user to establish a connection between test and hint is by the order of them in the resulting array and this breaks it. Maybe we shouldn't do that? Also, see the general review result I'll add on this there.

Reply:
Sorting was performed to sort the test results by the number of required subtests for the test. But I agree that it is better to leave a link between the input and output data.